### ISSUE TITLE
[Feature Request] Add Filesystem cache

### ISSUE THREAD

**Issue by vesper8** [GoogleChrome/rendertron#338]
- Add a filesystem-cache provider, similar to datastore-cache or memory-cache, but persisting the cached render results to the file system.
- Allow `filesystem` to be a value for the `cache` option
- Allow configuration of the number of entries (can be hardcoded for now) that the cache can hold
- Add tests
- Add docs

------ Previous discussion ------

Currently the only cache options are in-memory with a hardcoded limit of 100 entries or using the Google Cloud datastore. I would really like it if there were more options, such as using the local filesystem for storing pre-rendered html based on urls.. this should be just as fast as in-memory cache and allow you to cache a much greater number of entries for sites that have huge amounts of pages such as mine with 40k pages.

It would be equally great if an API was available for invalidating, recaching or verifying if a url is already cached. It would also be great if we could configure objects in this filesystem cache to expire automatically after a period of time

Here's to hoping someone feels like taking this on this issue! Thanks

**Comment by AVGP** [comment]
Hey @vesper8, thanks for opening an issue.

I think you're actually having like three feature requests, if not four in this one:

1. Make the size of the cache a config option
2. Add a filesystem cache provider
3. An API endpoint to recache
4. API to invalidate or verify a URL

Note: There's also #37 for an endpoint to clear the cache. Would you be okay if I split this issue up like in the enumeration?

**Comment by AVGP** [comment]
Also: The datastore & memory cache should give you an idea on how to implement new cache providers, so if you wanna give it a go, I'd be happy to provide guidance!

**Comment by vesper8** [comment]
Yes by all means split it as you see fit, it's true that 3 of 4 would also benefit the other cache providers

I will see about giving this a try, the thing is I have not adopted typescript yet, nor have I forked a pure JS project before.. but maybe it's time I did

Thanks!

**Comment by vesper8** [comment]
In the meantime I've setup a datastore which was pleasantly easy enough.. there was a bit of confusion because Datastore is being replaced by Firestore and I wasn't sure whether to use the Native Mode or the Datastore Mode.. I chose the Datastore Mode and everything seems to be working good.

@AVGP Considering this filesystem alternative might take some time.. can we at least put priority on making the `const cacheDurationMinutes = 60 * 24;` from https://github.com/GoogleChrome/rendertron/blob/master/src/datastore-cache.ts configurable via the `config.json` ?

I don't know why anyone thought hardcoding this very short expiry was a good idea... but I have 40k pages I'd like to have my search crawlers index easily and they all take about 3 seconds to render.. so I would much rather have them all cached with a very long expiry.. or ideally.. cached forever and have an API endpoint for invalidating/recaching them.

This should be an easy PR? I'm just not sure how it should look on the config.json

I would have thought that it would be nice if it were all under a "cache" key.. such as 
```
{
    "cache": {
        "provider": "datastore",
        "expiry_in_minutes": 10080
    }
}
```

But then this wouldn't be backwards compatible with the current configuration object which has `"cache": "datastore"` on the top level

So then adding more keys such as `"cache_expiry_in_minutes"` ? What do you think?

**Comment by AVGP** [comment]
> I don't know why anyone thought hardcoding this very short expiry was a good idea...

Because having *some* caching is better than having none, b/c whoever wrote that might not have known how / had the time to implement this as a config setting.

We are about to produce a new major version anyway (other things introduced backwards compatibility breaking changes), so I'd say now is a good time to restructure the config, like you suggested.

I filed #339 for the config part, I'll add more issues for the other parts.

**Comment by vesper8** [comment]
> > I don't know why anyone thought hardcoding this very short expiry was a good idea...
> 
> Because having _some_ caching is better than having none, b/c whoever wrote that might not have known how / had the time to implement this as a config setting.

Right.. apologies for my tone there.. thank you for breaking this up and taking it on, really excited to see future enhancements to this most excellent project. I tried out prerender.io (or tried to) and after having it fail on me over and over, I tried Rendertron instead and it was a joy from the beginning and so easy to set up in comparison. It just needs a little more control over the caching options and it will be top notch!
